azure-ts-dynamicresource/cdnCustomDomain.ts

```diff
@@ -0,0 +1,258 @@
+}
+
+/**
+ * DynamicProviderInputs represents the inputs that are passed as inputs
+ * to each function in the implementation of a `pulumi.dynamic.ResourceProvider`.
+ * The property names in this must match `CustomDomainOptions`.
+ */
+interface DynamicProviderInputs {
+    resourceGroupName: string;
+    profileName: string;
+    endpointName: string;
+    customDomainHostName: string;
+    httpsEnabled: boolean;
+}
+
+/**
+ * DynamicProviderOutputs represents the output type of `create` function in the
+ * dynamic resource provider.
+ */
+interface DynamicProviderOutputs extends DynamicProviderInputs, cdnManagement.CdnManagementModels.CustomDomainsCreateResponse {
+    name: string;
+}
+
+class CDNCustomDomainResourceProvider implements pulumi.dynamic.ResourceProvider {
+    private name: string;
+
+    constructor(name: string) {
+        this.name = name;
+    }
+
+    private async getCDNManagementClient(): Promise<cdnManagement.CdnManagementClient> {
+        let clientID = azure.config.clientId;
+        let clientSecret = azure.config.clientSecret;
+        let tenantID = azure.config.tenantId;
+        let subscriptionID = azure.config.subscriptionId;
+        let credentials: ServiceClientCredentials;
+
+        // If at least one of them is empty, try looking at the env vars.
+        if (!clientID || !clientSecret || !tenantID || !subscriptionID) {
+            console.log("Checking env vars for ARM credentials.");
+            clientID = process.env["ARM_CLIENT_ID"];
+            clientSecret = process.env["ARM_CLIENT_SECRET"];
+            tenantID = process.env["ARM_TENANT_ID"];
+            subscriptionID = process.env["ARM_SUBSCRIPTION_ID"];
+        }
+
+        // If they are still empty, try to get the creds from Az CLI.
+        if (!clientID || !clientSecret || !tenantID || !subscriptionID) {
+            // `create()` will throw an error if the Az CLI is not installed or `az login` has never been run.
+            const cliCredentials = await msRestAzure.AzureCliCredentials.create();
+            subscriptionID = cliCredentials.subscriptionInfo.id;
+            credentials = cliCredentials;
+        } else {
+            credentials = await msRestAzure.loginWithServicePrincipalSecret(clientID, clientSecret, tenantID);
+        }
+
+        return new cdnManagement.CdnManagementClient(credentials, subscriptionID);
+    }
+
+    async check(olds: DynamicProviderInputs, news: DynamicProviderInputs): Promise<pulumi.dynamic.CheckResult> {
+        const cdnClient = await this.getCDNManagementClient();
+
+        // If none of the CDN properties changed, then there is nothing to be validated.
+        if (olds.profileName === news.profileName &&
+            olds.endpointName === news.endpointName &&
+            olds.customDomainHostName === news.customDomainHostName) {
+            return { inputs: news };
+        }
+
+        const validationOutput = await cdnClient.endpoints.validateCustomDomain(
```

Comment:
When I run preview on a clean stack, I get an error (probably, because the resource group doesn't exist yet):

```
    Error: Resource group '04da6b54-80e4-46f7-96ec-b56ff0331ba9' could not be found.: Error: Resource group '04da6b54-80e4-46f7-96ec-b56ff0331ba9' could not be found.
        at new RestError (C:\work\pulumi\examples\azure-ts-dynamicresource\node_modules\@azure\ms-rest-js\lib\restError.ts:18:5)
        at C:\work\pulumi\examples\azure-ts-dynamicresource\node_modules\@azure\ms-rest-js\lib\policies\deserializationPolicy.ts:117:27
        at process._tickCallback (internal/process/next_tick.js:68:7)
```

Reply:
Hmm..does it pass when you re-run it?

Reply:
Something is not quite right here. I am looking into this.

Reply:
I have updated the provider to address this. The cause of this error was because `check` seems to be called during a preview as well, when the resource group doesn't exist yet. That is unfortunate, because the call to the validation of the custom domain seemed appropriate to have in the `check` function. I have moved it to the `create` function now. The `check` function is now just validating that there is a value for the required inputs.

When I used this for my own project originally, it was an existing app to which I added the custom domain, so the resource group et al. existed already, which is why I didn't notice this.